docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,340 @@
+type IPAddressClaimStatus struct {
+   // Address is a reference to the address that was allocated for this claim.
+   Address LocalObjectReference `json:"address,omitempty"`
+
+  // Conditions provide details about the status of the claim.
+	// +optional
+	Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+}
+ 
+// IPAddressClaim can be used to allocate IPAddresses from an IP Pool.
+type IPAddressClaim struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+ 
+   Spec   IPAddressClaimSpec   `json:"spec,omitempty"`
+   Status IPAddressClaimStatus `json:"status,omitempty"`
+}
+```
+
+##### IPAddress
+
+IPAddress resources are created to fulfil an IPAddressClaim. They are created by the IPAM provider that reconciles the claim.
+
+IPAddresses are immutable.
+
+```go
+// IPAddressSpec describes an IPAddress
+type IPAddressSpec struct {
+   // Claim is a reference to the claim this IPAddress was created for.
+   Claim LocalObjectReference `json:"claim,omitempty"`
+ 
+   // Pool is a reference to the pool that this IPAddress was created from.
+   Pool LocalObjectReference `json:"pool,omitempty"`
+ 
+   // Address is the IP address.
+   Address string `json:"address"`
+ 
+   // Prefix is the prefix of the address.
+   Prefix int `json:"prefix,omitempty"`
+ 
+   // Gateway is the network gateway of network the address is from.
+   Gateway string `json:"gateway,omitempty"`
+}
+ 
+// IPAddress is a representation of an IP Address that was allocated from an IP Pool.
+type IPAddress struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+ 
+   Spec IPAddressSpec `json:"spec,omitempty"`
+}
+```
+
+##### New Reference Type
+
+```go
+type LocalObjectReference struct {
+  Group string
+  Kind string
+  Name string
+}
+```
+
+#### Additional Notes
+
+- An example provider that provides in-cluster IP Address management (also useful for testing, or when no other IPAM solution is available) should be implemented as a separate project in a separate repository.
+- A shared library for providers and consumers should be implemented as part of that in-cluster IP Address provider.
+  - helpers to create IPAddresses
+  - a predicate to filter IPAddressClaims
+  - ... anything else that can be shared between providers and consumers
+
+### Security Model
+
+- IPAddressClaim, IPAddress and Pools are required to be in the same namespace, and also in the same namespace as the Cluster.
+  - For IPAddressClaim and IPAddress this is enforced by the lack of a namespace field on the reference
+  - Infrastructure Providers need to ensure that only Pools in the same namespace can be referenced, ideally by excluding namespace from the reference.
+
+### Risks and Mitigations
+
+- Unresponsive IPAM providers can prevent successful creation of a cluster
+- Wrong IP address allocations (e.g. duplicates) can brick a cluster’s network
+
+## Alternatives
+
```

Comment:
I'm curious how does this proposal aligns with metal3-io/ip-address-manager cc @furkatgofurov7?

Reply:
If metal3 adds support for this proposal it could be discontinued in my opinion. It currently only provides in-cluster IP address management. But it was built in a way that would allow adding different providers as well. I've started working on an in-cluster provider for this proposal which is much simpler as there is no generic layer in between.

As per this proposal, integration with different IPAM solutions would be implemented as separate IPAM providers. We will start working on a provider for Infoblox as soon as this proposal is merged.

Since we're using metal3, we'd also be happy to contribute integration for this proposal there.

Reply:
> I'm curious how does this proposal aligns with metal3-io/ip-address-manager plans cc @furkatgofurov7?

Hey,
Great discussion has been going on here lately! We are considering the possibility of switching to the IPAM solution outlined in this proposal.

> If metal3 adds support for this proposal it could be discontinued in my opinion. It currently only provides in-cluster IP address management. But it was built in a way that would allow adding different providers as well. I've started working on an in-cluster provider for this proposal which is much simpler as there is no abstraction layer in between.
> 

@schrej Thanks, we are planning on discontinuing the use of our IPAM solution and going with this one but we would like to make sure it is fully compatible. 




> Since we're using metal3, we'd also be happy to contribute integration for this proposal there.

:+1: